### DISCUSSION TITLE
Weird permission issues because  PUID/PGID is being ignored I think?

### DISCUSSION THREAD

**Discussion by wesgould** [mbround18/valheim-docker#786] [Q&A]
I am running into weird permission issues if I try to pass in the PUID and PGID Environment variables to the container. It appears as though the PUID is being passed properly but the PGID might not be. 

Here is my compose file: 

```
version: "3"
services:
  valheim:
    image: mbround18/valheim:latest
    restart: always
    container_name: valheim
    stop_signal: SIGINT
    ports:
      - 2456:2456/udp
      - 2457:2457/udp
      - 2458:2458/udp
    environment:
      PORT: 2456
      NAME: ""
      WORLD: ""
      PASSWORD: ""
      TZ: "America/New_York"
      PUBLIC: 1
      AUTO_UPDATE: 1
      AUTO_UPDATE_SCHEDULE: "0 1 * * *"
      AUTO_BACKUP: 1
      AUTO_BACKUP_SCHEDULE: "*/15 * * * *"
      AUTO_BACKUP_REMOVE_OLD: 1
      AUTO_BACKUP_DAYS_TO_LIVE: 3
      AUTO_BACKUP_ON_UPDATE: 1
      AUTO_BACKUP_ON_SHUTDOWN: 1
      UPDATE_ON_STARTUP: 0
      PUID: "1001"
      PGID: "1001"
    volumes:
      - ./valheim/saves:/home/steam/.config/unity3d/IronGate/Valheim
      - ./valheim/server:/home/steam/valheim
      - ./valheim/backups:/home/steam/backups
```

My server has my personal user at 1000:1000 and this valheim user at 1001:1001

if I do docker-compose up (with or without --build)  it  starts to launch but I get lots of permission denied errors:

```
Creating network "new_valheim_default" with the default driver
Creating valheim ... done
Attaching to valheim
valheim    | America/New_York
valheim    | 2024-01-28 11:06:36 - #[INFO]: ###########################################################################
valheim    | 2024-01-28 11:06:36 - [Valheim][INFO]: Valheim Server - Sun Jan 28 11:06:36 AM EST 2024
valheim    | 2024-01-28 11:06:36 - [Valheim][INFO]: Initializing your container...
valheim    | 2024-01-28 11:06:36 - [Valheim][INFO]: Total memory: 31GB
valheim    | 2024-01-28 11:06:36 - #[INFO]: ###########################################################################
valheim    | Configuring Preset Env
valheim    | export DEBUG_MODE=0
valheim    | export ODIN_CONFIG_FILE="/home/steam/valheim/config.json"
valheim    | export ODIN_DISCORD_FILE="/home/steam/valheim/discord.json"
valheim    | export ODIN_WORKING_DIR=""
valheim    | export SAVE_LOCATION="/home/steam/.config/unity3d/IronGate/Valheim"
valheim    | export MODS_LOCATION="/home/steam/staging/mods"
valheim    | export GAME_LOCATION="/home/steam/valheim"
valheim    | export BACKUP_LOCATION="/home/steam/backups"
valheim    | export NAME=""
valheim    | export ADDRESS=""
valheim    | export PORT=2456
valheim    | export PUBLIC=1
valheim    | export ENABLE_CROSSPLAY=0
valheim    | export UPDATE_ON_STARTUP=0
valheim    | export SERVER_EXTRA_LAUNCH_ARGS=""
valheim    | export WEBHOOK_URL=""
valheim    | export WEBHOOK_STATUS_SUCCESSFUL=1
valheim    | export WEBHOOK_STATUS_FAILED=1
valheim    | export WEBHOOK_STATUS_RUNNING=1
valheim    | export WEBHOOK_INCLUDE_PUBLIC_IP=0
valheim    | export AUTO_UPDATE=1
valheim    | export AUTO_BACKUP=1
valheim    | export AUTO_BACKUP_REMOVE_OLD=1
valheim    | export AUTO_BACKUP_DAYS_TO_LIVE=3
valheim    | export AUTO_BACKUP_ON_UPDATE=1
valheim    | export AUTO_BACKUP_ON_SHUTDOWN=1
valheim    | export AUTO_BACKUP_PAUSE_WITH_NO_PLAYERS=0
valheim    | export VALHEIM_PLUS_RELEASES_URL=""
valheim    | export VALHEIM_PLUS_DOWNLOAD_URL=""
valheim    | export BEPINEX_RELEASES_URL="https://valheim.thunderstore.io/api/experimental/package/denikson/BepInExPack_Valheim/"
valheim    | export BEPINEX_DOWNLOAD_URL=""
valheim    | export BEPINEX_FULL_RELEASES_URL=""
valheim    | Preset Env Configured
valheim    | 2024-01-28 11:06:37 - [Valheim][INFO]: Auto Update Enabled...
valheim    | 2024-01-28 11:06:37 - [Valheim][INFO]: Auto Update Schedule: 0 1 * * *
valheim    | mkdir: cannot create directory ‘/home/steam/valheim/logs’: Permission denied
valheim    | 0 1 * * * BASH_ENV=/env.sh /bin/bash /home/steam/scripts/auto_update.sh >> /home/steam/valheim/logs/auto-update.out 2>&1
valheim    | 2024-01-28 11:06:37 - [Valheim][INFO]: Auto Backup Enabled...
valheim    | 2024-01-28 11:06:37 - [Valheim][INFO]: Auto Backup Schedule: */15 * * * *
valheim    | mkdir: cannot create directory ‘/home/steam/valheim/logs’: Permission denied
valheim    | */15 * * * * BASH_ENV=/env.sh /bin/bash /home/steam/scripts/auto_backup.sh >> /home/steam/valheim/logs/auto-backup.out 2>&1
valheim    | 2024-01-28 11:06:37 - [Valheim][INFO]: Setting up file systems
valheim    | mkdir: cannot create directory ‘/home/steam/valheim/logs’: Permission denied
valheim    | usermod: no changes
valheim    | /entrypoint.sh: line 151: cd: /home/steam: Permission denied
valheim    | /entrypoint.sh: line 168: log: command not found
valheim    | /entrypoint.sh: line 177: line: command not found
valheim    | Configuring Preset Env
valheim    | Preset Env Configured
valheim    | /entrypoint.sh: line 187: log: command not found
valheim    | /entrypoint.sh: line 188: log: command not found
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | tee: /home/steam/cron.d//auto-update: Permission denied
valheim    | 0 1 * * * BASH_ENV=/env.sh /bin/bash /home/steam/scripts/auto_update.sh >> /home/steam/valheim/logs/auto-update.out 2>&1
valheim    | tee: /home/steam/cron.d//auto-update: Permission denied
valheim    | chmod: cannot access '/home/steam/cron.d//auto-update': Permission denied
valheim    | /entrypoint.sh: line 198: log: command not found
valheim    | /entrypoint.sh: line 199: log: command not found
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | tee: /home/steam/cron.d//auto-backup: Permission denied
valheim    | */15 * * * * BASH_ENV=/env.sh /bin/bash /home/steam/scripts/auto_backup.sh >> /home/steam/valheim/logs/auto-backup.out 2>&1tee: /home/steam/cron.d//auto-backup: Permission denied
valheim    | 
valheim    | chmod: cannot access '/home/steam/cron.d//auto-backup': Permission denied
valheim    | cat: '/home/steam/cron.d/*': Permission denied
valheim    | /entrypoint.sh: line 125: log: command not found
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | usermod: no changes
valheim    | /entrypoint.sh: line 151: cd: /home/steam: Permission denied
valheim exited with code 1
valheim    | /entrypoint.sh: line 168: log: command not found
valheim    | /entrypoint.sh: line 177: line: command not found
valheim    | Configuring Preset Env
valheim    | Preset Env Configured
valheim    | /entrypoint.sh: line 187: log: command not found
valheim    | /entrypoint.sh: line 188: log: command not found
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | tee: /home/steam/cron.d//auto-update: Permission denied
valheim    | 0 1 * * * BASH_ENV=/env.sh /bin/bash /home/steam/scripts/auto_update.sh >> /home/steam/valheim/logs/auto-update.out 2>&1
valheim    | tee: /home/steam/cron.d//auto-update: Permission denied
valheim    | chmod: cannot access '/home/steam/cron.d//auto-update': Permission denied
valheim    | /entrypoint.sh: line 198: log: command not found
valheim    | /entrypoint.sh: line 199: log: command not found
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | tee: /home/steam/cron.d//auto-backup: Permission denied
valheim    | */15 * * * * BASH_ENV=/env.sh /bin/bash /home/steam/scripts/auto_backup.sh >> /home/steam/valheim/logs/auto-backup.out 2>&1tee: /home/steam/cron.d//auto-backup: Permission denied
valheim    | 
valheim    | chmod: cannot access '/home/steam/cron.d//auto-backup': Permission denied
valheim    | cat: '/home/steam/cron.d/*': Permission denied
valheim    | /entrypoint.sh: line 125: log: command not found
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | mkdir: cannot create directory ‘/home/steam’: Permission denied
valheim    | /entrypoint.sh: line 151: cd: /home/steam: Permission denied
valheim    | usermod: no changes
valheim exited with code 1
```

If I run this, with my valheim (1001:1001) user  without any PUID or PGID args it creates a structure where the directory and files are all 1000:1000 aka my regular user account owns them. BUT IT WORKS*. I just want all of these files/directories and everything to be owned by my valheim 'service' account that I created for additional security. 

I think the steam user inside the container is running at 1000:1000 by default and it seems to not like it if you pass the variables.
```
valheim    | 2024-01-28 11:22:31 - #[INFO]: ###########################################################################
valheim    | 2024-01-28 11:22:31 - [Valheim][INFO]: Valheim Server - Sun Jan 28 11:22:31 AM EST 2024
valheim    | 2024-01-28 11:22:31 - [Valheim][INFO]: STEAM_UID 1000 - STEAM_GUID 1000
valheim    | 2024-01-28 11:22:31 - [Valheim][INFO]: Installing Valheim via odin 2.1.0...
valheim    | 2024-01-28 11:22:31 - #[INFO]: ###########################################################################
``` 

What am I doing wrong? I am sure its something super dumb on my end, but maybe its a bug? 

For what its worth I search for an answer, saw this https://github.com/mbround18/valheim-docker/pull/18  and https://github.com/mbround18/valheim-docker/issues/23 but nothing in there seemed to point me in the right direction other than it seems like it was fixed years ago?

**Comment by wesgould** [comment]
I guess it does look hardcoded to 1000:1000

Is this intentional? Is there a reason I shouldn't/couldn't change this to 1001:1001 for my case? 

https://github.com/mbround18/valheim-docker/blob/main/src/scripts/entrypoint.sh

```
setup_filesystem() {
  log "Setting up file systems"
  STEAM_UID=${PUID:=1000}
  STEAM_GID=${PGID:=1000}
```

**Reply by mbround18** [reply]
@wesgould The given bash commands STEAM_UID=${PUID:=1000} and STEAM_GID=${PGID:=1000} are setting default values for STEAM_UID and STEAM_GID variables if they aren't already set. This can be expanded into if statements as follows:


```bash
if [ -z "${PUID}" ]; then
 PUID=1000
fi

if [ -z "${PGID}" ]; then
 PGID=1000
fi

STEAM_UID=$PUID
STEAM_GID=$PGID
```

If you would like to learn more about bash expansion read here: https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html. its a short hand way of writing

**Reply by mbround18** [reply]
However, i can look into the shell and enforce that its looking for the correct variables

**Reply by mbround18** [reply]
Looks like when i changed the current user from root to steam it blocks the usermod logic for updating the user ids. looking into a fix rn

**Reply by mbround18** [reply]
@wesgould can you test https://github.com/mbround18/valheim-docker/pull/788 with the container `mbround18/valheim:sha-3361ed1`?

**Reply by mbround18** [reply]
right nowi have it update the group so its accessible to steam at 1000 and the group at what ever your number is

**Comment by wesgould** [comment]
So that seems to work, but I was trying avoid having my user account used at all (since it has elevated privileges).

```
drwxr-xr-x 2 wes     valheim 4096 Jan 28 19:15 backups/
drwxr-xr-x 3 wes     valheim 4096 Jan 28 19:02 saves/
drwxr-xr-x 7 wes     valheim 4096 Jan 28 19:02 server/
```
Ideally it would be valheim valheim since valheim is my non-privileged account just for this container. 

What would happen if you hardcoded the PUID=1001? Does it "steam" have to be 1000 to work?

**Reply by mbround18** [reply]
in this current configuration i cannot change the puid, is it possible you can add your user id to the valheim group?

**Reply by mbround18** [reply]
@wesgould wanted to check in

**Reply by wesgould** [reply]
I'd welcome anyone smarter than me re: security best practices, but my understanding is that one of the best practices for docker containers is having a separate user (on the host) for each container--where those users are limited in what they can do (eg non-privileged accounts/can't read or write or execute any other files on the system). If an exploit or bug allowed access to the host from the docker, that access would be from whatever permissions the 'user' uid has to write to the host. 

With that said, I dont think the solution above solves/meets the intent of that. So while it gets rid of the error, it still makes uid 1000 the owner of the files, and I think it most people's cases thats going to be their main user account that likely has sudo privileges. 

I'd be happy to help dig into the code to see where the dependencies on uid 1000 are. There should be a way to make the container write to the host as a different UID/GID than from whatever the container uses internally.

This is the most performant valheim docker server so I'd love to help fix that.

**Reply by mbround18** [reply]
Hey @wesgould, i have a background & degree in adv cyber security and wanted to chine in with this. The container was running as root which is user 0 in the container. This is seen as not a good practice and so the user was moved to the steam user.

I change this to be the steam user which is at `1000`, the steam user cannot change its own user ID in linux which is why the group was implemented.  If you have a user with `1002` they should also have a group of `1002` under that users name. if you have a system with multiple users per container just set the group permission to match that users permission and they should have access to those files under the hosts file system. 


Here is the best practices guides: https://docs.docker.com/develop/security-best-practices/ 
Security best practices: https://medium.com/@maheshwar.ramkrushna/best-practices-for-secure-docker-containerization-non-root-user-read-only-volumes-and-resource-d34ed09b1bd3
Stackoverflow example: https://stackoverflow.com/questions/70520205/docker-non-root-user-best-practices-for-python-images
Redhats article: https://developers.redhat.com/articles/2023/03/23/10-tips-writing-secure-maintainable-dockerfiles#10_tips_and_best_practices_for_dockerfiles
Dev.to: https://dev.to/izackv/running-a-docker-container-with-a-custom-non-root-user-syncing-host-and-container-permissions-26mb


If you truly want isolation and security for your containers and docker host through unique users via a user per container. This is what you are talking about and how to do user namespace remapping: https://docs.docker.com/engine/security/userns-remap/

Ideally, if you want to enforce a specific user ID without having to remap user the docker features and best practices, the container itself is very lightweight and can be extended via:  https://www.phind.com/search?cache=ftjumt3pf9k4r7n6at6n0bue

```Dockerfile
FROM mbround18/valheim:latest
ARG USER_ID=1001
ARG GROUP_ID=1001
USER steam
USER root
RUN usermod -u $USER_ID steam && groupmod -g $GROUP_ID steam
RUN chown steam:steam /home/steam
USER steam
```
